src/main/java/build/buildfarm/instance/AbstractServerInstance.java

```diff
@@ -286,28 +286,31 @@ private void validateActionInputDirectory(
 
       Digest directoryDigest = directoryNode.getDigest();
       if (!visited.contains(directoryDigest)) {
-        path.push(directoryDigest);
-        validateActionInputDirectory(expectDirectory(directoryDigest), path, visited, inputDigests);
-        path.pop();
-        visited.add(directoryDigest);
+        validateActionInputDirectory(directoryDigest, path, visited, inputDigests);
       }
     }
   }
 
-  private void validateActionInputs(Digest inputRootDigest, ImmutableSet.Builder<Digest> inputDigests) {
-    Stack<Digest> path = new Stack<>();
-    path.push(inputRootDigest);
-
-    Directory root = expectDirectory(inputRootDigest);
-    validateActionInputDirectory(root, path, new HashSet<>(), inputDigests);
+  private void validateActionInputDirectory(
+      Digest directoryDigest,
+      Stack<Digest> path,
+      Set<Digest> visited,
+      ImmutableSet.Builder<Digest> inputDigests) {
+    if (directoryDigest.getSizeBytes() != 0 || Directory.getDefaultInstance().toByteString().size() != 0) {
```

Comment:
Wait, an empty dir digest means an empty directory, because Directory.getDefaultInstance().toByteString() serializes to an empty string. And if it didn't, I don't quite understand this condition -- wouldn't it mean an empty digest is not a valid directory? Also, don't you want to add the empty digest to visited?
Also, I think it be safer to optimize for empty digests as close to the upload/download as possible -- in the getBlob function, for instance, instead of finding every place that uses downloads directly or indirectly. That way ensures things don't break when anyone anywhere asks to get an empty blob from CAS.

Reply:
Yeah, this is unnecessarily complicated, and was a hedge for the first run for me against getDefaultInstance() serializing to an empty string. If an empty string was not a serialization of the default instance, I wanted it to be proven invalid by the Directory.parseFrom. I'll simplifiy it and make sure that we're safe for upload/download and that we guard/optimize on either side of the interface.